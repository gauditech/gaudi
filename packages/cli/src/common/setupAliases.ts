import path from "path";

import { addAliases } from "module-alias";

/**
 * NOTE: Make sure to keep aliases in sync with `tsconfig.json` and `jest.config.js`.
 */

addAliases({
  "@cli": path.join(__dirname, "src"),
});
