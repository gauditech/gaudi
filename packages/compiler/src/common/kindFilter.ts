type Kind = string | number | symbol | boolean | undefined | null;

export type FilteredByKind<T extends { kind: Kind }, kind = T["kind"]> = kind extends T["kind"]
  ? Exclude<{ [K in keyof T]: K extends "kind" ? kind : T[K] }, { kind: never }>
  : never;

export type RejectedByKind<T extends { kind: Kind }, kind = T["kind"]> = Exclude<
  FilteredByKind<T>,
  { kind: kind }
>;

export function kindFilter<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): FilteredByKind<i[number], k>[] {
  return input.filter((i) => {
    for (const kind of kinds) {
      if (kind === i.kind) return true;
    }
    return false;
  }) as FilteredByKind<i[number], k>[];
}

export function kindReject<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): RejectedByKind<i[number], k>[] {
  return input.filter((i) => {
    for (const kind of kinds) {
      if (kind !== i.kind) return true;
    }
    return false;
  }) as RejectedByKind<i[number], k>[];
}

export function kindFind<i extends { kind: Kind }[], const k extends i[number]["kind"]>(
  input: i,
  ...kinds: k[]
): FilteredByKind<i[number], k> | undefined {
  return input.find((i) => {
    for (const kind of kinds) {
      if (kind === i.kind) return true;
    }
    return false;
  }) as FilteredByKind<i[number], k> | undefined;
}
