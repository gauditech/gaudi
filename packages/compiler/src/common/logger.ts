import { AsyncLocalStorage } from "node:async_hooks";
import util from "util";

import chalk from "chalk";
import _ from "lodash";
import { createLogger, format, transports } from "winston";

const als = new AsyncLocalStorage<string[]>();

/**
 * Logger implementation
 */

const winstonLogger = createLogger({
  level: "silly",
  format: format.combine(
    format.timestamp({
      format: "YYYY-MM-DD HH:mm:ss",
    }),
    format.label(),
    format.errors({ stack: true }),
    format.splat(),
    format.align(),
    format.padLevels(),
    format.printf(({ level, label, message, timestamp, stack }) => {
      const cStack = coloredStack(stack);
      const cTimestamp = coloredTs(timestamp);
      const cLevel = coloredLevel(level);
      const cLabel = coloredLabel(label);
      const msg = stack ? "" : padMessage(message);

      return `${cTimestamp} ${cLevel}${cLabel} ${msg}${cStack}`;
    })
  ),
  transports: [new transports.Console({})],
});

type LogLevel = "error" | "warn" | "info" | "verbose" | "debug" | "silly";

interface LogConfig {
  [category: string]: LogLevel;
}

const LOG_CATEGORIES: LogConfig = {
  "": "silly",
  sql: "debug",
  http: "debug",
  debug: "silly",
};

type Loggable = unknown;

export class Logger {
  private _category: string | null = null;
  private _forceLevel: LogLevel | undefined;

  constructor(category?: string, level?: LogLevel) {
    this._category = category ?? null;
    this._forceLevel = level;
  }

  silly(label: Loggable, message: Loggable): void;
  silly(message: Loggable): void;
  silly(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("silly", args[1], args[0]);
    } else {
      // only message is passed
      this._log("silly", args[0], undefined);
    }
  }

  debug(label: Loggable, message: Loggable): void;
  debug(message: Loggable): void;
  debug(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("debug", args[1], args[0]);
    } else {
      // only message is passed
      this._log("debug", args[0], undefined);
    }
  }

  info(label: Loggable, message: Loggable): void;
  info(message: Loggable): void;
  info(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("info", args[1], args[0]);
    } else {
      // only message is passed
      this._log("info", args[0], undefined);
    }
  }

  error(label: Loggable, message: Loggable): void;
  error(message: Loggable): void;
  error(...args: Loggable[]) {
    if (args.length > 1) {
      this._log("error", args[1], args[0]);
    } else {
      // only message is passed
      this._log("error", args[0], undefined);
    }
  }

  wrap<X>(name: string, cb: () => X): X {
    // NOTE: We don't currently implement anything that can utilize `als`
    const ctx = als.getStore();
    const newCtx = [...(ctx ?? []), name];
    return als.run(newCtx, cb) as X;
  }

  static specific(name: string, level?: LogLevel) {
    return new Logger(name, level);
  }

  private _log(level: LogLevel, message: Loggable, label: Loggable | undefined) {
    const minLogLevel = this._forceLevel ?? LOG_CATEGORIES[this._category || ""] ?? "silly";
    if (Logger.compareLevels(level, minLogLevel)) {
      this._do_log(level, message, label);
    }
  }

  protected _do_log(level: LogLevel, message: Loggable, label: Loggable | undefined): void {
    const msgObj =
      message instanceof Error || typeof message === "string"
        ? message
        : util.inspect(message, { depth: 10, colors: true });
    winstonLogger.log(level, msgObj as any, { label });
  }

  private static compareLevels(a: LogLevel, b: LogLevel): boolean {
    const levels: LogLevel[] = ["error", "warn", "info", "verbose", "debug", "silly"];
    return levels.indexOf(a) - levels.indexOf(b) <= 0;
  }
}

const logger = new Logger();
export default logger;

/**
 * Helper functions.
 */

function coloredLevel(level: string): string {
  const l = chalk.bold(chalk.white(` ${level.toUpperCase().padEnd(5)} `));
  switch (level) {
    case "silly": {
      return chalk.bgYellow(l);
    }
    case "debug": {
      return chalk.bgBlue(l);
    }
    case "info": {
      return chalk.bgGreen(l);
    }
    case "error": {
      return chalk.bgRed(l);
    }
    default: {
      return level;
    }
  }
}

function coloredTs(ts: string): string {
  return chalk.bold(`[${ts}]`);
}

function coloredStack(stack: string): string {
  if (!stack) {
    return "";
  }
  const [err, ...s] = stack.split("\n").map((line) => `${_.repeat("\t", 2)}${line}`);
  return `${chalk.bold(chalk.red(err))}\n${chalk.gray(s.join("\n"))}`;
}

function coloredLabel(label: string | undefined): string {
  return label ? chalk.bgGray(chalk.yellowBright(chalk.bold(`[${label}]`))) : "";
}

function padMessage(msg: string): string {
  return msg
    .split("\n")
    .map((line, index) => (index > 0 ? `\t\t${line}` : line))
    .join("\n");
}
