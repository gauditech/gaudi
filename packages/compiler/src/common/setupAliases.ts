import path from "path";

import { addAliases } from "module-alias";

/**
 * NOTE: Make sure to keep aliases in sync with `tsconfig.json` and `jest.config.js`.
 */

addAliases({
  "@compiler": path.join(__dirname, ".."),
});
