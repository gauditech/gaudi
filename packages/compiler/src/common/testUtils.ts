import { compileToAST } from "@compiler/compiler";
import { compilerErrorsToString } from "@compiler/compiler/compilerError";
import { migrate } from "@compiler/compiler/migrate";
import { compose } from "@compiler/composer/composer";
import { Definition } from "@compiler/types/definition";

/**
 * Helper function that compiles directly to definition. This is used in tests.
 */
export function compileFromString(source: string): Definition {
  const inputs = [{ source }];
  const { ast, errors } = compileToAST([{ source }]);
  if (errors) {
    const errorString = compilerErrorsToString(inputs, errors);
    throw new Error(`Failed to compile:\n${errorString}`);
  }
  return compose(migrate(ast));
}
