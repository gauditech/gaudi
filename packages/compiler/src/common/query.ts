import { assertUnreachable, ensureEqual } from "@compiler/common/utils";
import { SelectDef, SelectItem, TypedExprDef } from "@compiler/types/definition";
import _ from "lodash";
import { match } from "ts-pattern";

export function transformSelectPath(select: SelectDef, from: string[], to: string[]): SelectDef {
  return select.map((item: SelectItem): SelectItem => {
    return match(item)
      .with({ kind: "expression" }, (item) => ({
        ...item,
        expr: transformExpressionPaths(item.expr, from, to),
      }))
      .with({ kind: "model-hook" }, (item) => ({
        ...item,
        namePath: transformNamePath(item.namePath, from, to),
      }))
      .with({ kind: "nested-select" }, (item) => {
        return {
          ...item,
          namePath: transformNamePath(item.namePath, from, to),
          select: transformSelectPath(item.select, from, to),
        };
      })
      .exhaustive();
  });
}

export function transformNamePath(path: string[], from: string[], to: string[]): string[] {
  ensureEqual(
    _.isEqual(from, _.take(path, from.length)),
    true,
    `Cannot transform name path: ${path.join(".")} doesn't start with ${from.join(".")}`
  );
  return [...to, ..._.drop(path, from.length)];
}

export function transformNamePaths(paths: string[][], from: string[], to: string[]): string[][] {
  return paths.map((path) => transformNamePath(path, from, to));
}

export function transformExpressionPaths(
  exp: TypedExprDef,
  from: string[],
  to: string[]
): TypedExprDef {
  if (exp === undefined) {
    return undefined;
  }
  switch (exp.kind) {
    case "literal":
    case "variable": {
      return exp;
    }
    case "alias": {
      return { ...exp, namePath: transformNamePath(exp.namePath, from, to) };
    }
    case "function": {
      return {
        ...exp,
        args: exp.args.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    case "aggregate-function": {
      return { ...exp, sourcePath: transformNamePath(exp.sourcePath, from, to) };
    }
    case "in-subquery": {
      return {
        ...exp,
        sourcePath: transformNamePath(exp.sourcePath, from, to),
        lookupExpression: transformExpressionPaths(exp, from, to),
      };
    }
    case "array": {
      return {
        ...exp,
        elements: exp.elements.map((arg) => transformExpressionPaths(arg, from, to)),
      };
    }
    default: {
      assertUnreachable(exp);
    }
  }
}
