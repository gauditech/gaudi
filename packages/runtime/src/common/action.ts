import { transformSelectPath } from "@gaudi/compiler/dist/common/query";
import { dataToFieldDbnames, getRef } from "@gaudi/compiler/dist/common/refs";
import { assertUnreachable } from "@gaudi/compiler/dist/common/utils";
import {
  ActionDef,
  ChangesetDef,
  CreateOneAction,
  Definition,
  UpdateOneAction,
} from "@gaudi/compiler/dist/types/definition";
import _ from "lodash";

import { applyFilterIdInContext, buildQueryTree, queryTreeFromParts } from "../query/build";
import { castToCardinality, createQueryExecutor, executeQueryTree } from "../query/exec";

import { buildChangeset } from "@runtime/common/changeset";
import { executeActionHook } from "@runtime/hooks";
import { RequestContext, Storage } from "@runtime/server/context";
import { DbConn } from "@runtime/server/dbConn";
import { HookError } from "@runtime/server/error";

export async function executeEndpointActions(
  def: Definition,
  dbConn: DbConn,
  reqCtx: RequestContext,
  actions: ActionDef[]
) {
  //
  return _internalExecuteActions(def, dbConn, reqCtx, actions);
}
export async function executeActions(
  def: Definition,
  dbConn: DbConn,
  reqCtx: RequestContext,
  actions: ActionDef[]
) {
  return _internalExecuteActions(def, dbConn, reqCtx, actions);
}
async function _internalExecuteActions(
  def: Definition,
  dbConn: DbConn,
  reqCtx: RequestContext,
  actions: ActionDef[]
) {
  const qx = createQueryExecutor(dbConn);

  for (const action of actions) {
    const actionKind = action.kind;
    if (actionKind === "create-one") {
      const model = getRef.model(def, action.model);
      const dbModel = model.dbname;

      const actionChangeset = await buildChangeset(def, qx, reqCtx, action.changeset);
      const dbData = dataToFieldDbnames(model, actionChangeset);

      const id = await insertData(dbConn, dbModel, dbData);
      const deps = await fetchActionDeps(def, dbConn, action, id);
      deps && reqCtx.set(["aliases", action.alias], findOne(deps));
    } else if (actionKind === "update-one") {
      const model = getRef.model(def, action.model);
      const dbModel = model.dbname;

      const actionChangeset = await buildChangeset(def, qx, reqCtx, action.changeset);
      const dbData = dataToFieldDbnames(model, actionChangeset);

      const targetId = resolveTargetId(reqCtx, action.targetPath);

      const id = await updateData(dbConn, dbModel, dbData, targetId);
      const deps = await fetchActionDeps(def, dbConn, action, id);
      deps && reqCtx.set(["aliases", action.alias], findOne(deps));
    } else if (actionKind === "delete-one") {
      const model = getRef.model(def, action.model);
      const dbModel = model.dbname;

      const targetId = resolveTargetId(reqCtx, action.targetPath);

      await deleteData(dbConn, dbModel, targetId);
    } else if (actionKind === "query") {
      const qt = buildQueryTree(def, action.query);
      // FIXME this ugly
      const fieldset = reqCtx.get("fieldset") as Record<string, unknown>;
      const varsObj = Object.fromEntries(
        Object.keys(fieldset).map((k) => [`___changeset___${k}`, fieldset[k]])
      );
      varsObj["___requestAuthToken"] = reqCtx._express.req.user?.token;

      const result = castToCardinality(
        await qx.executeQueryTree(def, qt, new Storage(varsObj), []),
        action.query.retCardinality
      );
      reqCtx.set(["aliases", action.alias], result);
    } else if (actionKind === "execute-hook") {
      const argsChangeset = await buildChangeset(def, qx, reqCtx, action.hook.args);
      const hookCtx = { request: reqCtx._express.req, response: reqCtx._express.res };

      try {
        const result = await executeActionHook(def, action.hook.hook, argsChangeset, hookCtx);
        reqCtx.set(["aliases", action.alias], result);
      } catch (err) {
        throw new HookError(err);
      }
    } else if (actionKind == "respond") {
      try {
        // construct changeset and resolve action parts
        const actionChangesetDef: ChangesetDef = _.compact([
          { kind: "basic", name: "body", setter: action.body },
          action.httpStatus != null
            ? { kind: "basic", name: "httpStatus", setter: action.httpStatus }
            : undefined,
        ]);
        const changeset = await buildChangeset(def, qx, reqCtx, actionChangesetDef);
        // http headers are a separate changeset
        const httpHeadersChangesetDef: ChangesetDef = (action.httpHeaders ?? []).map((h) => ({
          kind: "basic",
          name: h.name,
          setter: h.value,
        }));
        const httpHeadersChangeset = await buildChangeset(def, qx, reqCtx, httpHeadersChangesetDef);

        const body = changeset.body;
        // we're forcing number cause our type system should've made sure that this resolves to appropriate type
        const httpResponseCode = (changeset.httpStatus ?? 200) as number;

        Object.entries(httpHeadersChangeset).forEach(([name, value]) => {
          // null - remove current header
          if (value == null) {
            reqCtx._express.res.removeHeader(name);
          }
          // multiple header values
          else if (_.isArray(value)) {
            reqCtx._express.res.set(name, value);
          }
          // single value
          else {
            // we're forcing value to `any` cause our type system should've made sure that this resolves to appropriate type
            reqCtx._express.res.set(name, value as any);
          }
        });
        reqCtx._express.res.status(httpResponseCode).json(body);
      } catch (err: any) {
        throw new Error(err);
      }
    } else if (actionKind === "validate") {
      throw new Error("Not implemented");
    } else {
      assertUnreachable(actionKind);
    }
  }
}

/**
 * Fetches the record targeted by action based on the actions `select` deps.
 */
async function fetchActionDeps(
  def: Definition,
  dbConn: DbConn,
  action: CreateOneAction | UpdateOneAction,
  id: number
): Promise<Record<string, unknown>[] | undefined> {
  if (!action.alias) return;
  // no need to fetch if only ID is requested
  if (action.select.findIndex((item) => item.alias !== "id") < 0) {
    return [{ id }];
  }
  const qt = queryTreeFromParts(
    def,
    action.alias,
    [action.model],
    applyFilterIdInContext([action.model]),
    transformSelectPath(action.select, [action.alias], [action.model])
  );
  return executeQueryTree(dbConn, def, qt, new Storage({}), [id]);
}

function resolveTargetId(reqCtx: RequestContext, targetPath: string[]): number {
  // append "id" to the end of the target path and separate it from it's root
  // eg.
  // [org, repo, issue] -> [org, repo, issue, id] -> [org, [repo, issue, id]]
  const [rootTargetName, ...path] = [...targetPath, "id"];

  return reqCtx.get([rootTargetName, ...path]) as number;
}

// ---------- DB functions

async function updateData(
  dbConn: DbConn,
  model: string,
  data: Record<string, unknown>,
  targetId: number
): Promise<number> {
  // avoid malformed query by skipping the update if no data is passed
  if (Object.keys(data).length === 0) {
    return targetId;
  }
  const ret = await dbConn(model).update(data).where({ id: targetId }).returning("id");
  return findOne(ret).id;
}

async function insertData(
  dbConn: DbConn,
  model: string,
  data: Record<string, unknown>
): Promise<number> {
  const ret = await dbConn.insert(data).into(model).returning("id");

  return findOne(ret).id;
}

async function deleteData(dbConn: DbConn, model: string, targetId: number): Promise<number> {
  const ret = await dbConn(model).delete().where({ id: targetId }).returning("id");

  return findOne(ret).id;
}

function findOne<T>(rows: T[], message?: string): T {
  if (rows.length === 0) {
    throw new Error(message ?? `Record not found`);
  }
  if (rows.length > 1) {
    throw new Error(`Unexpected error: multiple records found`);
  }

  return rows[0];
}
