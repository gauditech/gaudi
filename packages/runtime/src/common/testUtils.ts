import { compileToAST } from "@gaudi/compiler/dist/compiler";
import { compilerErrorsToString } from "@gaudi/compiler/dist/compiler/compilerError";
import { migrate } from "@gaudi/compiler/dist/compiler/migrate";
import { compose } from "@gaudi/compiler/dist/composer/composer";
import {
  CustomManyEndpointDef,
  Definition,
  QueryAction,
  QueryDef,
} from "@gaudi/compiler/dist/types/definition";

import { QueryTree } from "@runtime/query/build";
import { NestedRow, QueryExecutor } from "@runtime/query/exec";
import { Storage } from "@runtime/server/context";

/**
 * Helper function that compiles directly to definition. This is used in tests.
 */
export function compileFromString(source: string): Definition {
  const inputs = [{ source }];
  const { ast, errors } = compileToAST([{ source }]);
  if (errors) {
    const errorString = compilerErrorsToString(inputs, errors);
    throw new Error(`Failed to compile:\n${errorString}`);
  }
  return compose(migrate(ast));
}

/**
 * Creates dummy query executor wich always return empty row.
 *
 * TODO: add abbility to define dummy return data
 */
export function mockQueryExecutor(): QueryExecutor {
  return {
    executeQuery(
      _def: Definition,
      _q: QueryDef,
      _ctx: Storage,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },

    executeQueryTree(
      _def: Definition,
      _qt: QueryTree,
      _ctx: Storage,
      _contextIds: number[]
    ): Promise<NestedRow[]> {
      return Promise.resolve([]);
    },
  };
}
/**
 * Provide models blueprint, and query blueprint separately.
 * This function will compile a blueprint and extract the query definition,
 * so you can use it in tests.
 */

export function makeTestQuery(models: string, query: string): { def: Definition; query: QueryDef } {
  const bp = `
  ${models}

  model TestHelperModel {}
  api {
    entrypoint TestHelperModel {
      custom endpoint {
        method GET
        cardinality many
        path "/test"
        action {
          ${query}
        }
      }
    }
  }
  `;
  const def = compileFromString(bp);
  const endpoint = def.apis[0].entrypoints[0].endpoints[0] as CustomManyEndpointDef;
  const action = endpoint.actions[0] as QueryAction;
  return { def, query: action.query };
}
