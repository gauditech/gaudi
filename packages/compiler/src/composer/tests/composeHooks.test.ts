import { compileFromString } from "@compiler/common/testUtils";
import {
  CreateEndpointDef,
  CreateOneAction,
  CustomOneEndpointDef,
} from "@compiler/types/definition";

describe("compose hooks", () => {
  it("composes source hooks", () => {
    const bp = `
      runtime MyRuntime {
        source path "some/path/to/file"
      }

      validator Test {
        arg name { type string }
        assert hook {
          arg name name
          runtime MyRuntime
          source someHook from "githubc.js"
        }
        error { code "test" }
      }

      model Org {
        field name { type string, validate { Test() } }
        hook description {
          runtime MyRuntime
          source someHook from "githubc.js"
        }
      }

      api {
        entrypoint Org {
          create endpoint {
            action {
              create {
                set name hook {
                  runtime MyRuntime
                  source someHook from "hooks.js"
                }
              }
            }
          }
        }
      }

    `;
    const def = compileFromString(bp);
    const validatorHook = def.validators.at(-1)?.assert;
    const modelHook = def.models[0].hooks[0];
    const actionHook = (
      (def.apis[0].entrypoints[0].endpoints[0] as CreateEndpointDef).actions[0] as CreateOneAction
    ).changeset.find(({ name }) => name === "name");

    expect({ validatorHook, modelHook, actionHook }).toMatchSnapshot();
  });

  it("inline hooks", () => {
    const bp = `
      runtime MyRuntime {
        source path "some/path/to/file"
      }

      validator Test {
        arg name { type string }
        assert hook {
          arg name name
          inline "name === 'name'"
        }
        error { code "test" }
      }

      model Org {
        field name { type string, validate { Test() } }
        hook description {
          runtime MyRuntime
          inline "'some description'"
        }
      }

      api {
        entrypoint Org {
          create endpoint {
            action {
              create {
                set name hook {
                  inline "'test name'"
                }
              }
            }
          }
        }
      }

    `;
    const def = compileFromString(bp);
    const validatorHook = def.validators.at(-1)?.assert;
    const modelHook = def.models[0].hooks[0];
    const actionHook = (
      (def.apis[0].entrypoints[0].endpoints[0] as CreateEndpointDef).actions[0] as CreateOneAction
    ).changeset.find(({ name }) => name === "name");

    expect({ validatorHook, modelHook, actionHook }).toMatchSnapshot();
  });

  it("defaults to the default execution runtime when hook runtime is empty", () => {
    const bp = `
      runtime MyRuntime {
        default
        source path "some/path/to/file"
      }

      runtime MyRuntime2 {
        source path "some/path/to/file"
      }

      model Org { field name { type string } }

      api {
        entrypoint Org {
          create endpoint {
            action {
              create {
                set name hook {
                  source randomSlug from "hooks.js"
                }
              }
            }
          }
        }
      }

    `;

    const result = compileFromString(bp);

    expect(result.apis[0].entrypoints[0].endpoints[0]).toMatchSnapshot();
  });

  it("action hook", () => {
    const bp = `
      runtime MyRuntime {
        source path "some/path/to/file"
      }

      model Org {
        field name { type string }
      }

      api {
        entrypoint Org as org {

          custom endpoint {
            path "somePath"
            method POST
            cardinality one

            extra inputs {
              // test action inputs
              field termsOfUse { type boolean }
            }

            action {
              execute {
                hook {
                  // test hook args
                  arg name org.name
                  arg terms termsOfUse

                  runtime MyRuntime
                  source someHook from "hooks.js"
                }
              }
            }
          }
        }
      }
    `;
    const result = compileFromString(bp);

    expect(result.apis[0].entrypoints[0].endpoints).toMatchSnapshot();
  });

  it("composes action hook", () => {
    const bp = `
      model Org { field name { type string} }

      api {
        entrypoint Org {

          // login
          custom endpoint {
            path "somePath"
            method POST
            cardinality one

            extra inputs {
              field prop { type string }
            }

            action {
              execute {
                hook {
                  // action arg hook
                  arg user query { from Org, filter { id is 1 }, select { id, name }} // TODO: read from ctx - id

                  runtime @GAUDI_INTERNAL
                  source login from "hooks/auth"
                }
              }
            }
          }
        }
      }
    `;
    const def = compileFromString(bp);
    const action = (def.apis[0].entrypoints[0].endpoints[0] as CustomOneEndpointDef).actions[0];

    expect(action).toMatchSnapshot();
  });
});
