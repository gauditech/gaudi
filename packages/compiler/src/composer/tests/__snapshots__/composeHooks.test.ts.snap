// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compose hooks action hook 1`] = `
[
  {
    "actions": [
      {
        "alias": "$action_0",
        "hook": {
          "args": [
            {
              "kind": "basic",
              "name": "name",
              "setter": {
                "kind": "alias-reference",
                "path": [
                  "org",
                  "name",
                ],
                "source": "aliases",
              },
            },
            {
              "kind": "basic",
              "name": "terms",
              "setter": {
                "kind": "alias-reference",
                "path": [
                  "termsOfUse",
                ],
                "source": "fieldset",
              },
            },
          ],
          "hook": {
            "file": "hooks.js",
            "kind": "source",
            "runtimeName": "MyRuntime",
            "target": "someHook",
          },
        },
        "kind": "execute-hook",
        "responds": false,
      },
    ],
    "authSelect": [],
    "authorize": undefined,
    "fieldset": {
      "kind": "record",
      "nullable": false,
      "record": {
        "termsOfUse": {
          "kind": "field",
          "nullable": false,
          "required": true,
          "type": "boolean",
          "validate": undefined,
        },
      },
    },
    "kind": "custom-one",
    "method": "POST",
    "parentContext": [],
    "path": "somePath",
    "responds": true,
    "response": undefined,
    "target": {
      "alias": "org",
      "identifyWith": {
        "paramName": "org_id",
        "path": [
          "id",
        ],
        "type": "integer",
      },
      "kind": "model",
      "name": "Org",
      "namePath": [
        "Org",
      ],
      "retType": "Org",
      "select": [
        {
          "alias": "name",
          "expr": {
            "kind": "identifier-path",
            "namePath": [
              "Org",
              "name",
            ],
          },
          "kind": "expression",
          "type": {
            "kind": "string",
            "nullable": false,
          },
        },
        {
          "alias": "id",
          "expr": {
            "kind": "identifier-path",
            "namePath": [
              "Org",
              "id",
            ],
          },
          "kind": "expression",
          "type": {
            "kind": "integer",
            "nullable": false,
          },
        },
      ],
    },
  },
]
`;

exports[`compose hooks composes action hook 1`] = `
[
  {
    "alias": "orgQ",
    "kind": "query",
    "model": "Org",
    "query": {
      "filter": {
        "args": [
          {
            "kind": "identifier-path",
            "namePath": [
              "Org",
              "id",
            ],
          },
          {
            "kind": "alias-reference",
            "path": [
              "org",
              "id",
            ],
            "source": "aliases",
          },
        ],
        "kind": "function",
        "name": "is",
      },
      "fromPath": [
        "Org",
      ],
      "kind": "query",
      "limit": undefined,
      "modelRefKey": "Org",
      "name": "$query",
      "offset": undefined,
      "orderBy": undefined,
      "refKey": "N/A",
      "retCardinality": "collection",
      "retType": "Org",
      "select": [
        {
          "alias": "id",
          "expr": {
            "kind": "identifier-path",
            "namePath": [
              "Org",
              "id",
            ],
          },
          "kind": "expression",
          "type": {
            "kind": "integer",
            "nullable": false,
          },
        },
        {
          "alias": "name",
          "expr": {
            "kind": "identifier-path",
            "namePath": [
              "Org",
              "name",
            ],
          },
          "kind": "expression",
          "type": {
            "kind": "string",
            "nullable": false,
          },
        },
      ],
    },
  },
  {
    "alias": "$action_1",
    "hook": {
      "args": [
        {
          "kind": "basic",
          "name": "org",
          "setter": {
            "kind": "alias-reference",
            "path": [
              "orgQ",
            ],
            "source": "aliases",
          },
        },
        {
          "kind": "basic",
          "name": "int",
          "setter": {
            "kind": "literal",
            "literal": {
              "kind": "integer",
              "value": 1,
            },
          },
        },
      ],
      "hook": {
        "file": "hooks/auth",
        "kind": "source",
        "runtimeName": "@GAUDI_INTERNAL",
        "target": "login",
      },
    },
    "kind": "execute-hook",
    "responds": false,
  },
]
`;

exports[`compose hooks composes source hooks 1`] = `
{
  "actionHook": {
    "kind": "basic",
    "name": "name",
    "setter": {
      "hook": {
        "args": [],
        "hook": {
          "file": "hooks.js",
          "kind": "source",
          "runtimeName": "MyRuntime",
          "target": "someHook",
        },
      },
      "kind": "hook",
    },
  },
  "modelHook": {
    "args": [],
    "hook": {
      "file": "githubc.js",
      "kind": "source",
      "runtimeName": "MyRuntime",
      "target": "someHook",
    },
    "kind": "model-hook",
    "name": "description",
    "refKey": "Org.description",
  },
  "validatorHook": {
    "hook": {
      "args": [
        {
          "kind": "basic",
          "name": "name",
          "setter": {
            "kind": "identifier-path",
            "namePath": [
              "name",
            ],
          },
        },
      ],
      "hook": {
        "file": "githubc.js",
        "kind": "source",
        "runtimeName": "MyRuntime",
        "target": "someHook",
      },
    },
    "kind": "hook",
  },
}
`;

exports[`compose hooks defaults to the default execution runtime when hook runtime is empty 1`] = `
{
  "actions": [
    {
      "alias": "$action_0",
      "changeset": [
        {
          "kind": "basic",
          "name": "name",
          "setter": {
            "hook": {
              "args": [],
              "hook": {
                "file": "hooks.js",
                "kind": "source",
                "runtimeName": "MyRuntime",
                "target": "randomSlug",
              },
            },
            "kind": "hook",
          },
        },
      ],
      "isPrimary": true,
      "kind": "create-one",
      "model": "Org",
      "select": [
        {
          "alias": "id",
          "expr": {
            "kind": "identifier-path",
            "namePath": [
              "$action_0",
              "id",
            ],
          },
          "kind": "expression",
          "type": {
            "kind": "integer",
            "nullable": false,
          },
        },
      ],
      "targetPath": [
        "Org",
      ],
    },
  ],
  "authSelect": [],
  "authorize": undefined,
  "fieldset": undefined,
  "kind": "create",
  "parentContext": [],
  "response": [
    {
      "alias": "id",
      "expr": {
        "kind": "identifier-path",
        "namePath": [
          "Org",
          "id",
        ],
      },
      "kind": "expression",
      "type": {
        "kind": "integer",
        "nullable": false,
      },
    },
    {
      "alias": "name",
      "expr": {
        "kind": "identifier-path",
        "namePath": [
          "Org",
          "name",
        ],
      },
      "kind": "expression",
      "type": {
        "kind": "string",
        "nullable": false,
      },
    },
  ],
  "target": {
    "alias": "$target_0",
    "kind": "model",
    "name": "Org",
    "namePath": [
      "Org",
    ],
    "retType": "Org",
    "select": [
      {
        "alias": "id",
        "expr": {
          "kind": "identifier-path",
          "namePath": [
            "Org",
            "id",
          ],
        },
        "kind": "expression",
        "type": {
          "kind": "integer",
          "nullable": false,
        },
      },
    ],
  },
}
`;

exports[`compose hooks inline hooks 1`] = `
{
  "actionHook": {
    "kind": "basic",
    "name": "name",
    "setter": {
      "hook": {
        "args": [],
        "hook": {
          "inline": "'test name'",
          "kind": "inline",
        },
      },
      "kind": "hook",
    },
  },
  "modelHook": {
    "args": [],
    "hook": {
      "inline": "'some description'",
      "kind": "inline",
    },
    "kind": "model-hook",
    "name": "description",
    "refKey": "Org.description",
  },
  "validatorHook": {
    "hook": {
      "args": [
        {
          "kind": "basic",
          "name": "name",
          "setter": {
            "kind": "identifier-path",
            "namePath": [
              "name",
            ],
          },
        },
      ],
      "hook": {
        "inline": "name === 'name'",
        "kind": "inline",
      },
    },
    "kind": "hook",
  },
}
`;
